### ISSUE TITLE
Should I correct the warning "creating a tensor from a list of numpy.ndarray is extremely slow" ?

### ISSUE THREAD

**Issue by Ca-ressemble-a-du-fake** [CorentinJ/Real-Time-Voice-Cloning#893]
Hi, 

During the training, I get the following warning constantly :
`/content/Real-Time-Voice-Cloning/synthesizer/synthesizer_dataset.py:84: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  ../torch/csrc/utils/tensor_new.cpp:201.)`

Should I follow the advice and change line 84 to `embeds = torch.tensor(np.array(embeds))` ? Or this was done on purpose and must not be changed ?

**Comment by blue-fish** [comment]
It's not intentional. I had to rewrite the dataloader for synthesizer training in #472, but didn't optimize the code since there were so many other things to be done.

Would you please fix it and benchmark the training speed before and after making the change?

**Comment by Ca-ressemble-a-du-fake** [comment]
I'll do it on my fork and report here before committing on master (it's my first contribution on github).

**Comment by Ca-ressemble-a-du-fake** [comment]
While doing this I came across this :
`| Epoch: 3/18 (207/288) | Loss: 0.5240 | 1.4 steps/s | Step: 15k | }Traceback (most recent call last):
  File "synthesizer_train.py", line 35, in <module>
    train(**vars(args))
  File "/content/Real-Time-Voice-Cloning/synthesizer/train.py", line 192, in train
    if np.isnan(grad_norm.cpu()):
RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.`

And now this error comes every time I launch the training. I tried to revert the change but the error still appears. So it may not be linked to the `np.array` conversion but to the GPU (since it degraded to T100 instead of V100 initially). So I have to solve this first before doing the correction mentioned above.

**Comment by blue-fish** [comment]
What happens if you try it locally with your own GPU?

**Comment by Ca-ressemble-a-du-fake** [comment]
On my GPU it worked like a charm but there were no noticeable performance improvement. Only the warning went away.

**Comment by blue-fish** [comment]
We should still fix it, to avoid confusing the user. Please submit a pull request.

**Comment by Ca-ressemble-a-du-fake** [comment]
For the records, I benchmarked both versions this morning with a batch size of 16 and a reduction factor of 7 with english_cleanners, on a P100 PCIE. I tested it only for the first epoch.

- *Original* version : 0.82 steps / s
- *Np Array "pre conversion*" version : 0.80 steps / s

So the fix only removes the warning, and does not bring performance boost. Anyway I'll submit a pull request.

**Comment by Ca-ressemble-a-du-fake** [comment]
I launched some basic benchmarks because I did not understand why moving to `np.array` did not lead to better performance overall (higher steps / s). So I measured the time to create the speaker embedding `embeds` as a list and as an array :

`embeds = [x[2] for x in batch]` VS `embeds = np.array([x[2] for x in batch])`

On average array creation is 5.5 times slower than list creation.

Then I also measured the time to convert `embeds` as a tensor.

`embeds = torch.tensor(embeds)`

On average conversion from array to tensor was 12 times quicker than conversion from list to tensor.

**Bottom line : it is worth it using array.**

But also, as the "steps / s" rate does not increase, it means the bottleneck lies somewhere else.
@blue-fish  Should I further continue finding the bottleneck and improving the performance ?

**Comment by Ca-ressemble-a-du-fake** [comment]
One more thing regarding the use of `np.array` instead of `list` that I've just noticed while testing it on a Tesla P100 (16GB) GPU. With `r=7` and a batch_size of 64, the "list version" stops training after 84 steps while the "array version" keeps training. Memory usage amounts to around 14,5 GB.

With the "list version" a batch size of 20 was near the maximum I could reach with a P100 and [reportedly](https://github.com/coqui-ai/TTS/blob/main/docs/source/faq.md#how-can-i-check-model-performance) model convergence is hard to reach with a batch size less than 32. So this is really cool.

**Bottom line : using array is really worth it**!

**Comment by blue-fish** [comment]
@Ca-ressemble-a-du-fake Thank you for sharing the observations, and submitting pull request #895 which resolves this issue. I expect there is some performance optimization possible, and I would appreciate any assistance you can provide in locating the bottlenecks.
